zip-0221.rst

```diff
@@ -0,0 +1,559 @@
+        start_height: int
+        end_height: int
+        count_shielded_txs: int # number of shielded transactions in block
+
+        @classmethod
+        def from_block(Z, block: ZcashBlock) -> 'ZcashMMRNode':
+            '''Create a leaf node from a block'''
+            return Z(
+                left_child=None,
+                right_child=None,
+                subtree_commitment=block.header_hash,
+                start_time=block.timestamp,
+                end_time=block.timestamp,
+                start_target=block.nBits,
+                end_target=block.nBits,
+                start_sapling_root=block.sapling_root,
+                end_sapling_root=block.sapling_root,
+                subtree_total_work=calculate_work(block.nBits),
+                start_height=block.height,
+                end_height=block.height,
+                count_shielded_txs=block_shielded_tx_count)
+
+        def serialize(self) -> bytes:
+            '''serializes a node'''
+            return (
+                self.subtree_commitment
+                + serialize_uint32(self.start_time)
+                + serialize_uint32(self.end_time)
+                + serialize_uint32(self.start_target)
+                + serialize_uint32(self.end_target)
+                + start_sapling_root
+                + end_sapling_root
+                + serialize_uint256(self.subtree_total_work)
+                + serialize_compact_uint(self.start_height)
+                + serialize_compact_uint(self.end_height)
+                + serialize_compact_uint(self.count_shielded_txs))
+
+
+    def make_parent(
+            left_child: ZcashMMRNode,
+            right_child: ZcashMMRNode) -> ZcashMMRNode:
+        return ZcashMMRNode(
+            left_child=left_child,
+            right_child=right_child,
+            subtree_commitment=H(left_child.serialize() + right_child.serialize()),
+            start_time=left_child.start_time,
+            end_time=right_child.end_time,
+            start_target=left_child.start_target,
+            end_target=left_child.end_target,
+            start_sapling_root=left_child.sapling_root,
+            end_sapling_root=right_child.sapling_root,
+            subtree_total_work=left_child.subtree_total_work + right_child.subtree_total_work,
+            start_height=left_child.start_height,
+            end_height=right_child.end_height,
+            count_shielded_txs=left_child.count_shield + right_child.count_shield)
+
+    def make_root_commitment(root: ZcashMMRNode) -> bytes:
+        '''Makes the root commitment for a blockheader'''
+        return H(root.serialize())
+
+Incremental push and pop (pseudocode)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With each new block ``B_n``, we append a new MMR leaf node corresponding to block ``B_(n-1)``. The ``append`` operation is detailed below in pseudocode (adapted from [#FlyClient]_):
+
+.. code-block:: python
+
+    def append(root: ZcashMMRNode, leaf: ZcashMMRNode) -> ZcashMMRNode:
```

Comment:
seems this append pseudo-code after appending to this tree:

                               (---11g---)     
                              /                  \        
                          ( 6 )                  \        
                         /        \                 \        
                      (2)       (5)            (9)        
                      /   \       /    \          /   \         
                   (0)  (1) (3)   (4)   (7)   (8)   

will not produce correct tree of this form:

                                    (--12g--)
                                   /                \
                              (---11g---)      \
                             /                  \        \
                           ( 6 )                \        \
                          /        \               \         \
                       (2)       (5)          (9)         \
                      /   \       /    \         /   \           \
                  (0)  (1) (3)   (4) (7)   (8)       (10)

rather, it will produce this:

                                 (----12g----)
                                /                      \
                              /                          \
                           ( 6 )                    (11g)
                          /        \                /         \
                       (2)       (5)          (9)         \
                      /   \       /    \        /   \           \
                  (0)  (1) (3)   (4)  (7)   (8)      (10)

(with 'g' marked virtual elements)

which will in turn make deletion algorithm working incorrectly 

while this state will autocorrect itself with next append, intermediate state is not suitable for deletion pseudo-code!

Reply:
Have no idea how to format those trees to be displayed and edited at the same time correctly, editing the comment will reveal the difference in structure

Reply:
did you check this with bitwise or boolan `not`? There is definitely a bug in the pseudocode. The line should read     `if not (n_leaves & (n_leaves - 1)):`. But might be more clear as `    if (n_leaves & (n_leaves - 1)) == 0:`

Reply:
@prestwich
I just do `if n_leaves & (n_leaves - 1) == 0 { ... }` in real Rust code (which will pass only for 2^N)

not sure what are rules for `int` -> `bool` conversion for a pseudo-code though!

Reply:
ah, this is generally pythonic. So all non-0 integers are truthy. I'll go into this in depth tomorrow and get back to you :+1:

Reply:
@therealyingtong do you model this code off another implementation?

Reply:
I tested this on some basic trees. It's not as elegant. I'll try to work out something better on the plane tomorrow

```
def get_peaks(node: ZcashMMRNode) -> List[ZcashMMRNode]:
    peaks: List[ZcashMMRNode] = []

    left_child = node.left_child
    right_child = node.right_child

    # find the number of leaves in the subtree
    left_leaves = left_child.latest_height - left_child.earliest_height + 1
    right_leaves = right_child.latest_height - right_child.earliest_height + 1

    if (left_leaves & (left_leaves - 1)) == 0:
        peaks.append(left_child)
    else:
        peaks.extend(get_peaks(left_child))

    if (right_leaves & (right_leaves - 1)) == 0:
        peaks.append(right_child)
    else:
        peaks.extend(get_peaks(right_child))

    return peaks

def append(root: ZcashMMRNode, leaf: ZcashMMRNode) -> ZcashMMRNode:
    '''Append a leaf to an existing tree, return the new tree root'''
    # recursively find a list of peaks in the current tree
    peaks: List[ZcashMMRNode] = get_peaks(root)
    merged: List[ZcashMMRNode] = []

    # Merge peaks from right to left. 
    # This will produce a list of peaks in reverse order
    current = leaf
    for peak in peaks[::-1]:
        current_leaves = current.latest_height - current.earliest_height + 1
        peak_leaves = peak.latest_height - peak.earliest_height + 1

        if current_leaves == peak_leaves:
            current = make_parent(peak, current)
        else:
            merged.append(current)
            current = peak

    # finally, bag the merged peaks
    return bag_peaks(merged[::-1])
```

Reply:
cool, I think it will fail in case of full tree (exactly 2^N leaves) but easy to fix

will try it on multiple cases

Reply:
@therealyingtong can you update this?

Reply:
Rust code based on this updated pseudocode idea works ok, thanks!